File: protocol/src/test/php/zfoophp/ProtocolManager.php
<?php

namespace zfoophp;

include_once 'IProtocolRegistration.php';
include_once 'packet/EmptyObject.php';
include_once 'packet/NormalObject.php';
include_once 'packet/ObjectA.php';
include_once 'packet/ObjectB.php';
include_once 'packet/SimpleObject.php';

class ProtocolManager
{
    private static array $protocols = array();
    private static array $protocolIdMap = array();

    public static function initProtocol(): void
    {
        self::$protocols[0] = new EmptyObjectRegistration();
        self::$protocolIdMap[EmptyObject::class] = 0;
        self::$protocols[101] = new NormalObjectRegistration();
        self::$protocolIdMap[NormalObject::class] = 101;
        self::$protocols[102] = new ObjectARegistration();
        self::$protocolIdMap[ObjectA::class] = 102;
        self::$protocols[103] = new ObjectBRegistration();
        self::$protocolIdMap[ObjectB::class] = 103;
        self::$protocols[104] = new SimpleObjectRegistration();
        self::$protocolIdMap[SimpleObject::class] = 104;;
    }

    public static function getProtocolId(mixed $clazz): int
    {
        return self::$protocolIdMap[$clazz];
    }

    public static function getProtocol(int $protocolId): IProtocolRegistration
    {
        return self::$protocols[$protocolId];
    }

    public static function write(ByteBuffer $buffer, mixed $packet): void
    {
        $protocolId = self::getProtocolId($packet::class);
        // write protocol id to buffer
        $buffer->writeShort($protocolId);
        // write packet
        self::getProtocol($protocolId)->write($buffer, $packet);

    }

    public static function read(ByteBuffer $buffer): mixed
    {
        $protocolId = $buffer->readShort();
        return self::getProtocol($protocolId)->read($buffer);
    }
}

File: protocol/src/test/swift/zfooswift/ProtocolManager.swift
import Foundation

class ProtocolManager {
    static var protocols = Dictionary<Int, IProtocolRegistration>()

    static func initProtocol() {
        // initProtocol
        protocols[0] = EmptyObjectRegistration()
        protocols[101] = NormalObjectRegistration()
        protocols[102] = ObjectARegistration()
        protocols[103] = ObjectBRegistration()
        protocols[104] = SimpleObjectRegistration()
    }
    
    static func getProtocol(_ protocolId: Int) -> IProtocolRegistration {
        return protocols[protocolId]!
    }
    
    static func write(_ buffer: ByteBuffer, _ packet: Any) {
        let p = packet as! IProtocol
        let protocolId = p.protocolId()
        let pro = getProtocol(protocolId)
        buffer.writeShort(Int16(protocolId))
        pro.write(buffer, p)
    }
    
    static func read(_ buffer: ByteBuffer) -> Any {
        let protocolId = buffer.readShort()
        let pro = getProtocol(Int(protocolId))
        return pro.read(buffer)
    }
}